### PR TITLE
Streamlining #asWord simplifications

### PR DESCRIPTION
**Pull request by PetarMax**
This PR streamlines some `#asWord`-related simplifications.

### CODE REVIEW

File: kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/bytes-simplification.k [hunk 1]
```diff
@@ -179,18 +179,6 @@ module BYTES-SIMPLIFICATION [symbolic]
...
-      requires 0 <=Int W andBool 0 <=Int X andBool X <Int 2 ^Int (W *Int 8)
-       andBool #asInteger(B) ==Int 0
-      [concrete(B, W), simplification]
-
-    rule [range-buf-zero-concat-extend]:
-      B1:Bytes +Bytes #buf(W:Int, X:Int) +Bytes B2 => #buf(lengthBytes(B1) +Int W, X) +Bytes B2
-      requires 0 <=Int W andBool 0 <=Int X andBool X <Int 2 ^Int (W *Int 8)
-       andBool #asInteger(B1) ==Int 0
-      [concrete(B1, W), simplification]
-
```
Comment (PetarMax) [inline kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/bytes-simplification.k:193]: These are not needed anymore with the introduction of `[asWord-concat-right]`.

File: kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/evm-int-simplification.k [hunk 2]
```diff
@@ -57,6 +57,14 @@ module EVM-INT-SIMPLIFICATION-COMMON
   // #asWord
   // ###########################################################################
 
+  rule [asWord-lb]: 0 <=Int #asWord(_WS)     => true [simplification, smt-lemma]
+  rule [asWord-ub]: #asWord(_WS) <Int pow256 => true [simplification, smt-lemma]
```
Comment (PetarMax) [inline kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/evm-int-simplification.k:61]: These have been moved here from `lemmas.k`, with added `smt-lemma` to enable more SMT reasoning around `#asWord`.

File: kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/lemmas.k [hunk 3]
```diff
@@ -61,9 +61,6 @@ module LEMMAS [symbolic]
 
     rule #isPrecompiledAccount(#newAddr(_, _), _) => false [simplification]
 
-    rule 0 <=Int #asWord(_WS)             => true [simplification]
-    rule         #asWord(_WS) <Int pow256 => true [simplification]
-
```
Comment (PetarMax) [inline kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/lemmas.k:66]: Moved to `evm-int-simplifications.k`

**Comment by palinatolmach** [comment]
Thanks @PetarMax! I can confirm that these changes fix branching that occured in these two tests in Kontrol:
```solidity
    function test_dynamic_byte_read_v1(bytes memory data, uint256 offset) public {
        vm.assume(offset < data.length);
        uint8 mydata = uint8(data[offset]);
        vm.assume(mydata < 3);
        assertTrue(mydata == 2 || mydata == 1 || mydata == 0);
    }

    function test_dynamic_byte_read_v2(bytes memory data, uint256 offset, bytes1 value) public {
        vm.assume(offset < data.length);
        data[offset] = value;
        assertTrue(data[offset] == value);
    }
```

But it seems that several booster tests are failing here in KEVM (https://github.com/runtimeverification/evm-semantics/actions/runs/7901094206/job/21564122413?pr=2301), e.g., the output for `erc20/hkg/transferFrom-failure-2-spec.k` is
```
PROOF FAILED: f0d4602717667cc05239620de33a8499dfb5c725aa4bcfa12a98ff90d1a94187
2 Failure nodes. (0 pending and 2 failing)

Failing nodes:

  Node id: 17
  Failure reason:
    Structural matching failed, the following cells failed individually (antecedent #Implies consequent):
    OUTPUT_CELL: ( b"\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01" #Implies b"\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00" )
  Path condition:
    { true #Equals 0 <Int VALUE:Int }

  Node id: 21
  Failure reason:
    Structural matching failed, the following cells failed individually (antecedent #Implies consequent):
    OUTPUT_CELL: ( b"\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01" #Implies b"\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00" )
  Path condition:
    ( { true #Equals VALUE:Int <=Int ALLOW:Int }
#And { true #Equals 0 <Int VALUE:Int } )
```

**Comment by ehildenb** [comment]
If this reduces the branching in a Kontrol test, then we can add that branch condition here as a unit simplification test as well. Please do so!

**Comment by PetarMax** [comment]
This is very strange. I have revisited the simplifications and minimised the difference wrt `master` with both of the tests still passing.

**Comment by palinatolmach** [comment]
Thank you @PetarMax! I'll add the unit test with the simplification that's needed for Kontrol examples, as @ehildenb requested.

**Comment by PetarMax** [comment]
> I added a test claim that illustrates the simplification that now happens in Kontrol ([19688bd](https://github.com/runtimeverification/evm-semantics/commit/19688bdfa3fba0205f433870fca2b1e327e3962f)):
> 
> ```
> claim  [buf-asWord-invert]:
>   <k> runLemma( 
>         #asWord ( #buf ( 63 , lengthBytes ( VV0_data_114b9705:Bytes ) ) )
>       ) => doneLemma(
>          ( lengthBytes ( VV0_data_114b9705:Bytes ) ) 
>       ) ... </k>
>     requires lengthBytes ( VV0_data_114b9705:Bytes ) <=Int 1073741824
> ```
> 
> Initially, I thought
> 
> ```
> claim  [buf-asWord-invert]:
>   <k> runLemma( 
>         #asWord ( #buf ( 63 , lengthBytes ( VV0_data_114b9705:Bytes ) ) ) <=Int maxUInt64
>       ) => doneLemma(
>          true 
>       ) ... </k>
>     requires lengthBytes ( VV0_data_114b9705:Bytes ) <=Int 1073741824
> ```
> 
> would pass too (since `maxUInt64` is 18446744073709551615, and `lengthBytes (...)` <= 1073741824 < 18446744073709551615), but it simplifies to
> 
> ```
> lengthBytes ( VV0_data_114b9705:Bytes ) <=Int maxUInt64 #Implies true
> ```
> 
> instead.

For me the test passes, I've pushed the changes now.

**Review by palinatolmach — APPROVED** [review]

**Pull request merged**